src/raqm.c

```diff
@@ -1126,3 +1136,372 @@ _raqm_u32_to_u8_index (raqm_t   *rq,
+                                       output);
+  return length;
+}
+
+static bool
+_allowed_grapheme_boundary(hb_codepoint_t lChar,
+                                       hb_codepoint_t rChar);
+
+static bool
+_in_Hangul_Syllable(hb_codepoint_t Char);
+
+/**
+ * raqm_index_to_position:
+ * @rq: a #raqm_t.
+ * @index: character index.
+ * @x: (out): output x position.
+ * @y: (out): output y position.
+ *
+ * Calculates the cursor position after the character at @index. If the
+ * character is right-to-left, then the cursor will be at the left of it,
+ * whereas if the character is left-to-right, then the cursor will be at the
+ * right of it.
+ *
+ * Return value:
+ * %true if the process was successful, %false otherwise.
+ *
+ * Since: 0.1
+ */
+bool
+raqm_index_to_position (raqm_t *rq,
+                        size_t *index,
+                        int *x,
+                        int *y)
+{
+  bool found = false;
+
+  /* We don't currently support multiline, so y is always 0 */
+  *y = 0;
+  *x = 0;
+
+  if (rq == NULL)
+    return false;
+
+  if (rq->flags & RAQM_FLAG_UTF8)
+    *index = _raqm_u8_to_u32_index (rq, *index);
+
+  if (*index >= rq->text_len)
+    return false;
+
+  RAQM_TEST ("\n");
+
+  while(*index < rq->text_len)
+  {
+    if (_allowed_grapheme_boundary(rq->text[*index], rq->text[*index+1]))
+      break;
+    else
+      *index += 1;
+  }
+
+  for (raqm_run_t *run = rq->runs; run != NULL && !found; run = run->next)
+  {
+    size_t len;
+    hb_glyph_info_t *info;
+    hb_glyph_position_t *position;
+
+    len = hb_buffer_get_length (run->buffer);
+    info = hb_buffer_get_glyph_infos (run->buffer, NULL);
+    position = hb_buffer_get_glyph_positions (run->buffer, NULL);
+
+
+    for (size_t i = 0; i < len && !found; i++)
+    {
+      uint32_t curr_cluster = info[i].cluster;
+      uint32_t next_cluster = curr_cluster;
+      *x += position[i].x_advance;
+
+      if (run->direction == HB_DIRECTION_LTR)
+        for (size_t j = i + 1; j < len && next_cluster == curr_cluster; j++)
+          next_cluster = info[j].cluster;
+      else
+        for (int j = i - 1; i != 0 && j >= 0 && next_cluster == curr_cluster; j--)
+          next_cluster = info[j].cluster;
+
+      if (next_cluster == curr_cluster)
+        next_cluster = run->pos + run->len;
+
+      if (*index < next_cluster && *index >= curr_cluster)
+      {
+        if (run->direction == HB_DIRECTION_RTL)
+          *x -= position[i].x_advance;
+        *index = curr_cluster;
+        found = true;
+      }
+    }
+  }
+
+  *index = _raqm_u32_to_u8_index(rq, *index);
+  RAQM_TEST ("The position is %d at index %ld\n",*x ,*index);
+  return true;
+}
+
+/**
+ * raqm_position_to_index:
+ * @rq: a #raqm_t.
+ * @x: x position.
+ * @y: y position.
+ * @index: (out): output character index.
+ *
+ * Convert from @x and @y position within text to logical position of character
+ * @index. If the position is outside the text, the last character is chosen as
+ * @index.
```

Comment:
The documentation is not clear, should be something like `Returns the @index of the character at @x and @y position…`